[perfetto] initial port of 43.1 (#37959)
microsoft/vcpkg: ports/perfetto/CMakeLists.txt

cmake_minimum_required(VERSION 3.8.0)
project(perfetto)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

include(CMakePackageConfigHelpers)

include_directories(sdk)

> add_library(perfetto sdk/perfetto.cc)

Reviewer: I think we still need
```suggestion
add_library(perfetto sdk/perfetto.cc)
target_include_directories(perfetto PUBLIC $<LINK_INTERFACE:include>)
```
~~~

Author: Feedback should all be addressed.

if (MSVC)
    target_compile_options(perfetto PRIVATE "/bigobj")
    target_compile_options(perfetto PRIVATE "/permissive-")
    target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)
endif (MSVC)

install(
    TARGETS perfetto
    EXPORT unofficial-perfetto-config
    LIBRARY DESTINATION lib
    ARCHIVE DESTINATION lib
)

install(FILES sdk/perfetto.h DESTINATION include/)

install(
    EXPORT unofficial-perfetto-config

>     NAMESPACE unofficial::

Reviewer: The suggested pattern is:
```suggestion
    NAMESPACE unofficial::perfetto::
```

> cmake_minimum_required(VERSION 3.8.0)
> project(perfetto)
> 
> set(CMAKE_CXX_STANDARD 17)
> set(CMAKE_CXX_STANDARD_REQUIRED ON)
> 
> include(CMakePackageConfigHelpers)
> 
> add_library(perfetto sdk/perfetto.cc sdk/perfetto.h)
> target_include_directories(perfetto PUBLIC $<INSTALL_INTERFACE:include>)
> 
> if (MSVC)
>     target_compile_options(perfetto PRIVATE "/bigobj")
>     target_compile_options(perfetto PRIVATE "/permissive-")
>     target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)
> endif (MSVC)
> 
> install(
>     TARGETS perfetto
>     EXPORT unofficial-perfetto-config
>     LIBRARY DESTINATION lib
>     ARCHIVE DESTINATION lib
> )
> 
> install(FILES sdk/perfetto.h DESTINATION include/)
> 
> install(
>     EXPORT unofficial-perfetto-config
>     NAMESPACE unofficial::perfetto::
>     DESTINATION share/unofficial-perfetto
> )

Reviewer 2: ```suggestion
# Unofficial perfetto CMakeLists.txt from https://github.com/google/perfetto/blob/v43.1/meson.build
cmake_minimum_required(VERSION 3.23)
project(perfetto LANGUAGES CXX)

add_library(perfetto)
target_compile_features(perfetto PRIVATE cxx_std_17)
target_sources(perfetto
    PRIVATE "sdk/perfetto.cc"
    PUBLIC FILE_SET HEADERS BASE_DIRS "sdk" FILES "sdk/perfetto.h"
)

set(THREADS_PREFER_PTHREAD_FLAG TRUE)
find_package(Threads REQUIRED)
target_link_libraries(perfetto PRIVATE Threads::Threads)

if(ANDROID)
    target_link_libraries(perfetto PRIVATE log)
endif(ANDROID)

if(WIN32)
    target_compile_options(perfetto PRIVATE "/bigobj")
    target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)
    target_link_libraries(perfetto PRIVATE ws2_32)
endif(WIN32)

if(MSVC)
    target_compile_options(perfetto PRIVATE "/permissive-")
endif(MSVC)

install(TARGETS perfetto
    EXPORT unofficial-perfetto-config
    FILE_SET HEADERS DESTINATION "include"
)

install(EXPORT unofficial-perfetto-config
    NAMESPACE unofficial::perfetto::
    DESTINATION "share/unofficial-perfetto"
)

install(FILES "protos/perfetto/trace/perfetto_trace.proto" DESTINATION "share/unofficial-perfetto")
```

# Unofficial perfetto CMakeLists.txt from https://github.com/google/perfetto/blob/v43.1/meson.build
cmake_minimum_required(VERSION 3.23)
project(perfetto LANGUAGES CXX)

add_library(perfetto)

> target_compile_features(perfetto PRIVATE cxx_std_17)

Author: Yes, see the [v33.1 release notes](https://github.com/google/perfetto/releases/tag/v33.1).

> Switched to a C++17-only project by removing C++11 opt-out. This completes the migration started in v31.0.

target_sources(perfetto
    PRIVATE "sdk/perfetto.cc"
    PUBLIC FILE_SET HEADERS BASE_DIRS "sdk" FILES "sdk/perfetto.h"
)

> set(THREADS_PREFER_PTHREAD_FLAG TRUE)

Reviewer 3: Should `PThreads` be a dependency (or feature) in the `vcpkg.json`?

Author: Seems like it, yes. I'll add it.

> find_package(Threads REQUIRED)

Author: I'm unsure what you mean here aside from what's asked in the other comment.

> Should PThreads be a dependency (or feature) in the vcpkg.json?

> find_package(Threads REQUIRED)

Reviewer 2: The lib `pthread` is a system library, it just requires the `find_package`, and it will only be used on NON-WINDOWS. The checks in upstream https://github.com/google/perfetto/blob/v43.1/include/perfetto/ext/base/thread_checker.h#L22-L24.

target_link_libraries(perfetto PRIVATE Threads::Threads)

if(ANDROID)
    target_link_libraries(perfetto PRIVATE log)
endif(ANDROID)

if(WIN32)
    target_compile_options(perfetto PRIVATE "/bigobj")

>     target_compile_definitions(perfetto PRIVATE WIN32_LEAN_AND_MEAN NOMINMAX)

Reviewer 3: Why are these preprocessor definitions needed? (`WIN32_LEAN_AND_MEAN`, `NOMINMAX`)

Author: `NOMINMAX` is needed to prevent the definition of macros which conflict with perfetto's use of `std::min` and `std::max`.

`WIN32_LEAN_AND_MEAN` is needed to compile both `winsock.h` and `ws2def.h` successfully.

```
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(103): warning C4005: 'AF_IPX': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(457): note: see previous definition of 'AF_IPX'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(147): warning C4005: 'AF_MAX': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(476): note: see previous definition of 'AF_MAX'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(187): warning C4005: 'SO_DONTLINGER': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(399): note: see previous definition of 'SO_DONTLINGER'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(240): error C2011: 'sockaddr': 'struct' type redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(482): note: see declaration of 'sockaddr'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(442): error C2143: syntax error: missing '}' before 'constant'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(442): error C2059: syntax error: 'constant'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(496): error C2143: syntax error: missing ';' before '}'
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(496): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(496): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\shared\ws2def.h(577): warning C4005: 'IN_CLASSA': macro redefinition
C:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\um\winsock.h(284): note: see previous definition of 'IN_CLASSA'

```

    target_link_libraries(perfetto PRIVATE ws2_32)
endif(WIN32)

if(MSVC)

>     target_compile_options(perfetto PRIVATE "/permissive-")

Reviewer 3: Why is this needed?

Author: To be clear that turns _off_ permissive mode (note the minus), and puts MSVC into standards-compliant mode.

I wasn't able to get it to fail without this flag as part of the automated build, but my recollection from previous usage is that it's needed for older MSVC versions  (2017?). 

It's in the [recommended set of flags](https://perfetto.dev/docs/instrumentation/tracing-sdk) for using the tracing compiler:

```cmake
# Enable standards-compliant mode when using the Visual Studio compiler.
if (MSVC)
  target_compile_options(example PRIVATE "/permissive-")
endif (MSVC)
```

Reviewer 3: If it's not strictly required we should remove this. Thanks for the explanation!
